[3779] Form engine for Preview 2.0

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/3779

Commit:
Revert encoding changes

aachavarria committed:
adding childrenMap to redux,
Adding iframe to editform panel,
WIP, opening iframe on embedded components

aachavarria committed:
WIP

aachavarria committed:
opening embedded component

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -548,7 +547,7 @@ export class ContentController {
     const currentModels = ContentController.models$.value;
 
-    post(GUEST_MODELS_RECEIVED, responseModels);
+    //post(GUEST_MODELS_RECEIVED, responseModels);
```
Leave this as it was. Do `childrenMap` as separate event.

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -220,9 +220,20 @@ const reducer = createReducer<GlobalState['preview']>({
+              index: 5,
+              fieldId: ['features_o']
+            }
+          ],
```
Revert. Don't commit hard coded test data.

aachavarria replied:
updated.

aachavarria committed:
wip

aachavarria committed:
adding embedded-legacy-form;
browseComponentsPanel fixes;
publishingPackages fixes;

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3779

 Conflicts:
	ui/app/src/modules/Preview/Tools/BrowseComponentsPanel.tsx
	ui/app/src/utils/hooks.ts

aachavarria committed:
hooks.ts, adding partial to return type;
preview.ts, cleaning query on setContentTypeFilter;
root.ts, removing duplicate code;
GlobalNav-main.js, Adding ecryptionTool route;

aachavarria committed:
common-api.js, changes to change modal width and height to 100%, adding use getTopLegacyWindow();
form-engine, adding renderComplete callback;
editFormPanel, changes to call form legacy embedded;

aachavarria committed:
common-api.js, comment startAuthLoop();
form-engine.js, adding aux to openChild, adding renderComplete callback;
host.js, removing debugger;
legacy_form, adding success and cancelled callbacks;
editFormPanel, removing old code;

aachavarria committed:
adding try catch

aachavarria committed:
removing refresh on legacy form cancelled;
guest.js, fixing issue with image upload;

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3779

aachavarria committed:
adding ModelHelper to helpers.ts;
object.ts adding setProperty;

aachavarria committed:
supporting editFormPanel with nested childs

aachavarria committed:
setting height to 100% on studio-ice-dialogs

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3779

 Conflicts:
	static-assets/components/cstudio-common/common-api.js
	static-assets/scripts/main.js

aachavarria committed:
adding missing getTopLegacyWindow()

aachavarria committed:
adding loading state to editFormPanel Dialog

aachavarria committed:
updating preview Legacy to use CrafterCMSNext setSiteCookie;

aachavarria committed:
revert window.top

aachavarria committed:
adding callback for embedded child form, adding loading state;

aachavarria committed:
creating generateIndex, to support register elements on move in nested components

aachavarria committed:
updating delete op to support nested components

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -9766,3 +9766,19 @@ if (window.top === window) {
+
+function getTopLegacyWindow(nextWindow) {
+  try {
+    if(nextWindow && nextWindow.IS_LEGACY_TOP_WINDOW) {
```
First of all, check if current window is the top window.

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -9766,3 +9766,19 @@ if (window.top === window) {
+      return nextWindow;
+    }
+    if(nextWindow) {
+      // if(nextWindow === window) return window;
```
x

rart commented on static-assets/components/cstudio-forms/forms-engine.js:
```diff
@@ -1755,8 +1765,14 @@ var CStudioForms = CStudioForms || function() {
           });
+          // if(CStudioAuthoring.InContextEdit.getIceCallback(editorId).renderComplete) {
+          //   CStudioAuthoring.InContextEdit.getIceCallback(editorId).renderComplete();
+          // }
```
x

rart commented on static-assets/components/cstudio-forms/forms-engine.js:
```diff
@@ -1710,12 +1710,22 @@ var CStudioForms = CStudioForms || function() {
                 const order = message.order != null ? message.order : null;
                 const contentType = message.contentType || parseDOM(FlattenerState[message.key]).querySelector('content-type').innerHTML;
+                let callback = message.callback || {};
+                if (callback.renderComplete) {
```
Check if renderComplete is a string. If someone else sends a function it should work too.
Don't reassign `callback`. Just renderComplete

rart commented on templates/web/legacy/embedded_legacy_form.ftl:
```diff
@@ -0,0 +1,115 @@
+  <script src="/studio/static-assets/yui/json/json-min.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
+  <script src="/studio/static-assets/yui/yahoo/yahoo-min.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
+  <script src="/studio/static-assets/libs/jquery/dist/jquery-3.4.1.min.js"></script>
+    <#include "/templates/web/common/page-fragments/studio-context.ftl" />
```
align/format

rart replied:
This is getting repeatedly marked as resolved without getting it done. Make sure you do this this time.

 Check after you commit that this shows up align here.

If it’s the auto-format. Turn it off temporarily.

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -51,14 +51,15 @@ export interface PagedEntityState<T = any> extends EntityState<T> {
   modelId: string;
   fieldId: string[];
-  index: number;
+  index: string;
```
It can be either `string | number`

rart commented on ui/app/src/modules/Preview/Tools/EditFormPanel.tsx:
```diff
@@ -52,6 +158,52 @@ export default function EditFormPanel() {
+        window.removeEventListener('message', callback, false)
+      };
+    } catch {
+      console.log('No supported yet.')
```
Not*

What is it that wouldn't be supported? When does this fail?

aachavarria replied:
I think it is not failing right now, it was just a precaution.

rart commented on ui/app/src/modules/Preview/Tools/EditFormPanel.tsx:
```diff
@@ -52,6 +158,52 @@ export default function EditFormPanel() {
+        } else if (e && e.data && e.data.type === EMBEDDED_LEGACY_CHILD_FORM_RENDERED && !path) {
+          setLoading(false);
+        }
+      };
```
Here, use rxjs. The `e && e.data && e.data.type` move to a filter operator and the type check moves to a switch inside a subscribe.

rart commented on ui/app/src/modules/Preview/Tools/EditFormPanel.tsx:
```diff
@@ -68,45 +220,42 @@ export default function EditFormPanel() {
+            classes={{ root: classes.loadingRoot }}
+          />
+        }
+        <iframe src={src} title='cstudio-embedded-legacy-form'
```
Double quote HTML attributes. Also, title should be human-readable descriptive text. It's not a technical ID.

rart commented on ui/app/src/modules/System/Publishing/Queue/PublishingQueue.tsx:
```diff
@@ -204,42 +205,31 @@ function PublishingQueue(props: PublishingQueueProps) {
-  }, [apiState, currentFilters, filters, getPackages, packages, siteId]);
+  useEffect(() => {
+    getPackages(siteId);
+  }, [currentFilters, siteId, getPackages]);
```
currentFilters is not a dependency of this effect

aachavarria replied:
it is not a dependency, but it needs to change everytime that the currentFilters changes

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -222,7 +223,11 @@ const reducer = createReducer<GlobalState['preview']>({
-            ...payload
+            ...payload.models
+          },
+          childrenMap: {
```
Separate action for children map

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -191,7 +195,7 @@ export function GuestProxy(props) {
           // Update attribute(s)
           // $el.attr('data-craftercms-index', newIndex);
```
What's up with these bits of commented code? (lines 129, 196, 201)

aachavarria replied:
there are your comments, Im going to delete them

rart reviewed (changes requested):
Is this branch up to date with the repo? Changes in ui/guest suggest this could be outdated and the PR is showing changes from your local branches

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -557,6 +556,7 @@ export class ContentController {
       Object.values(responseModels).forEach((model) =>
         this.computeChildren(model)
       );
+      post(GUEST_MODELS_RECEIVED, { models: responseModels, childrenMap: this.children });
```
Do `childrenMap` as separate event.

aachavarria committed:
Creating action/reducer SET_CHILDREN_MAP

aachavarria committed:
getTopLegacyWindow fixes

aachavarria committed:
editFormPanel: changing to messages to rxjs;

aachavarria committed:
deleting comments

aachavarria committed:
formEngine: checking callback.renderComplete;
align

aachavarria committed:
typo;
minor changes;

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3779

aachavarria committed:
adding buttons to edit form, and edit controller

aachavarria committed:
template-editor: adding cancelled callback, renderComplete callback, failure callback;
embedded_legacy_form: adding js-next-scripts.ftl, lang resources;
editFormPanel: adding template, controller;

aachavarria committed:
common-api: fixing issue on openPreview

aachavarria committed:
form-engine: removing iceWindowCallback.cancelled();
templateEditor: adding cancelled on !modified;
embedded-legacy-form: adding boostrap;

aachavarria committed:
Adding tabs, for quick change to form-edit/ controller edit/ template edit;

aachavarria committed:
template-editor: changing id to class, and adding scope;
embedded_legacy_form: adding changes to support 2 templates editor;
editFormPanel: changes on loading;

aachavarria committed:
formEngine: getCustomsCallbacks();
template editor: adding pendingChanges;
embeddedLEgacyForm: sending active tab on postMessage;
EditFormPanel: useCallback, adding pending changes;

aachavarria committed:
editFormPanel: adding filterBy, to get some 'loaded' or 'pending';
forms-engine: adding datum onChange to check for pending changes;

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -1871,18 +1871,18 @@ var nodeOpen = false,
               });
 
-              $modal.appendTo(window.top.document.body);
+              $modal.appendTo(getTopLegacyWindow().document.body);
```
Try to make a single call to getTopLegacyWindow per function - it can become a little expensive to get it multiple times per function now that is not a simple property accessor.

rart commented on static-assets/components/cstudio-forms/forms-engine.js:
```diff
@@ -1549,9 +1566,9 @@ var CStudioForms = CStudioForms || function() {
-                        if (iceWindowCallback && iceWindowCallback.cancelled) {
-                          iceWindowCallback.cancelled();
-                        }
+                        // if (iceWindowCallback && iceWindowCallback.cancelled) {
```
x

rart commented on static-assets/components/cstudio-forms/template-editor.js:
```diff
@@ -427,7 +427,7 @@ CStudioAuthoring.Module.requireModule(
-											}
+											};
 											// aceEditor.completers = [staticWordCompleter]
 											langTools.addCompleter(customCompleter);
```
x

rart commented on static-assets/components/cstudio-forms/template-editor.js:
```diff
@@ -472,16 +472,16 @@ CStudioAuthoring.Module.requireModule(
+											addButton.className = "addButtonVar btn btn-primary";
 											addButton.innerHTML = "Add Code";
-											addButton.className = "btn btn-primary";
+											//addButton.className = "btn btn-primary";
```
x

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -50,6 +50,20 @@ export function getPreviewURLFromPath(baseUrl: string, path: string) {
+  if (results == null)
+    return "";
+  else
+    return results[1];
```
as discussed used the queryString lib.

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -557,6 +557,7 @@ export class ContentController {
       Object.values(responseModels).forEach((model) =>
         this.computeChildren(model)
       );
+      post(SET_CHILDREN_MAP, this.children);
```
SET_CHILDREN_MAP => CHILDREN_MAP_UPDATE

rart reviewed (changes requested):
Please see if this is missing merged updates. I think there are changes from other PRs/braches here (e.g. what's BrowseComponentsPanel doing here?)

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3779

aachavarria committed:
adding pendingChanges cb to rte;

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3779

aachavarria committed:
PR review

aachavarria committed:
adding scoped topWindow on functions;

aachavarria committed:
creating EmbeddedLegacyEditors.tsx;
Toolbar Menu;

aachavarria committed:
deleting old code on toolbar.tsx

aachavarria committed:
adding usePreviewGuest hook;
adding getContentItem service;
adding PublishDialog;

aachavarria committed:
adding refresh after asve on legacyEditors;
adding schedule;
adding pluck;

aachavarria committed:
Removing IcePanel;

rart commented on static-assets/scripts/host.js:
```diff
@@ -567,8 +567,10 @@
 
     path = path.replace('//', '/');
 
+    console.log(path);
```
x

rart commented on static-assets/scripts/host.js:
```diff
@@ -623,6 +625,7 @@
   function selectStudioContent(site, url) {
     CStudioAuthoring.Service.lookupContentItem(site, url, {
       success: function (content) {
+        console.log('maybe yo');
```
x

rart commented on static-assets/scripts/host.js:
```diff
@@ -567,8 +567,10 @@
+    console.log(path);
     CStudioAuthoring.Service.lookupContentItem(CStudioAuthoringContext.site, path, {
       success: function (content) {
+        console.log('soy yo');
```
x

rart commented on templates/web/legacy/embedded_legacy_form.ftl:
```diff
@@ -0,0 +1,212 @@
+  <script src="/studio/static-assets/libs/bootstrap/js/bootstrap.min.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
+    <#include "/templates/web/common/js-next-scripts.ftl" />
+    <#-- Lang resources -->
+    <#assign path="/studio/static-assets/components/cstudio-common/resources/" />
```
align

rart commented on ui/app/src/modules/Preview/ToolBar.tsx:
```diff
@@ -178,9 +288,39 @@ export function AddressBar(props: AddressBarProps) {
+        open={Boolean(anchorEl)}
+        onClose={handleClose}
+      >
+        <MenuItem onClick={() => handleEdit('form')}>Edit</MenuItem>
```
Replace all hard coded labels for translations

rart commented on ui/app/src/modules/Preview/ToolBar.tsx:
```diff
@@ -124,13 +145,102 @@ export function AddressBar(props: AddressBarProps) {
   } = props;
   const noSiteSet = isBlank(site);
   const [internalUrl, setInternalUrl] = useState(url);
+  const [anchorEl, setAnchorEl] = useState(null);
```
Feels like too much responsibility for this component. Let’s discuss refactor.

rart commented on ui/app/src/modules/Preview/Tools/EditFormPanel.tsx:
```diff
@@ -33,14 +75,105 @@ function createBackHandler(dispatch) {
   };
 }
 
+function findParentModelId(modelId: string, childrenMap: LookupTable<Array<string>>, models: any) {
```
Should be moved to an útil module.
Eventually shared with guest

rart commented on ui/app/src/modules/Preview/Tools/EditFormPanel.tsx:
```diff
@@ -33,14 +75,105 @@ function createBackHandler(dispatch) {
+  }, [childrenMap, contentTypes, defaultSrc, model, models, path, selectedContentType, selectedId, site]);
+
+  function openDialog(type: string) {
+    setDialogConfig(
```
format

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -43,13 +45,20 @@ export function getPathFromPreviewURL(previewURL: string) {
 
+export function getQueryVariable(query: string, variable: string) {
+  let qs = parse(query);
+  return qs[variable] || null;
```
use ?? instead of || so that blank doesn’t resolve to null

aachavarria committed:
removing console.log

aachavarria committed:
format

aachavarria committed:
PR Review

rart committed:
Update ToolBar.tsx

rart committed:
Update EditFormPanel.tsx
